[feat] 에뮬레이터-로컬-저장소-설계-및-구현2

tomatozil opened the pull request:
## 🔧 어떤 기능인가요?

- 에뮬레이터 전역으로 쓰이는 클래스와 저장소를 구현했습니다.
- repository 인터페이스와 그 구현체 InMemoryStore, 구현체가 의존하고 있는 static 클래스가 있습니다.
- Emulator에는 다음 필드들이 있습니다.
    - 차량번호: mdn, 
    - 터미널 아이디: tid, 
    - 제조사 아이디: mid, 
    - 패킷 버전: pv, 
    - 디바이스 아이디: did,
    - 시동 on 데이터(최근 on시 전달했던 데이터들): onGPS
    - 시동 off 데이터(최근 off시 전달했던 데이터들): offGPS
    - 누적 주행 거리: sum,
    - 제어 설정 정보: CtrList 클래스
        - 제어 설정 고유 id: ctrId,
        - 제어 설정 코드('05' 고정): ctrCd,
        - 제어 값(주기정보 전송 주기값 60, 120, 180. . .): ctrVal
    - 지오펜스 정보: GeoList 클래스
        - 지오펜스 고유 id: geoCtrId 
        - 지오펜스 업데이트 값: geoGrpId
        - 지오펜스 이벤트 타입(반경 진입시 알림, 반경 이탈시 알림, 반경 진입/이탈 모든 경우 알림): geoEvtTp
        - 지오펜스 반경: geoRange
        - 지오펜스 (중심) 위도: lat
        - 지오펜스 (중심) 경도: lon
        - 지오펜스 설정 시작 시간: onTime
        - 지오펜스 설정 끝 시간: offTime
        - 지오펜스 저장 타입: storeTp

- 고정 값은 Constant 클래스를 만들어서 따로 관리할 수 있도록 했습니다.
- repository 의 InMemoryStore 구현체는 이전에 저장된 데이터를 새로운 데이터로 갱신하는 것입니다.
- CtrList와 GeoList는 변경할 때 객체 자체를 갈아끼우도록 builder 패턴을 적용했습니다.

## #️⃣ 연관된 이슈

#3 

## 💡 리뷰어에게 하고 싶은 말

- builder 패턴을 일부 적용했는데, 괜찮을까요? 다른 모듈과 통일 할 필요는 없을까요?
- setter를 일부 적용했습니다. setter 사용을 지양하고 객체에게 직접 질의해야한다는 ~oop 원칙~이 있습니다만 여기에서 setter 적용한 것들은 값을 단순히 변경하는거라 괜찮다고 판단했습니다. 헉 그런데 생각해보니까 검증 코드가 들어간다면.. setter를 안쓰는게 맞겠군요. 정적 팩토리 메서드로 가야할 것 같습니다.

<!-- 코드에 나타나지 않는 설계 및 의도, 중점적으로 봐줬으면 하는 점, 특이한 변경사항 등 -->

## 🙏 아래 내용이 모두 충족 되었는지 확인해주세요 🙏

- [x] PR 이전 `dev` 브랜치 병합 하셨나요?
- [x] PR 이전 빌드 테스트 정상적으로 성공했나요?
- [x] PR 상세내용이 충분히 기재 되었나요?
- [x] PR 리뷰어, 할당자, 라벨, 프로젝트 확인

tomatozil committed:
feat: 에뮬레이터-로컬-저장소-설계-및-구현2

Jeongjjuna commented on monicar-emulator/src/main/java/org/emulator/infrastructure/Emulator.java:
```diff
@@ -0,0 +1,45 @@
+    @Setter
+    private String offGPS;
+    @Setter
+    private String sum;
```
setter 를 처음부터 열어두기보다는 구현하다가 수정이 필요한 부분이 있다면 의미있는 이름의 메서드로 작성해주는 것은 어떨까요?!

Jeongjjuna commented on monicar-emulator/src/main/java/org/emulator/application/EmulatorRepository.java:
```diff
@@ -0,0 +1,12 @@
+import org.emulator.infrastructure.GeoList;
+
+public interface EmulatorRepository {
+    void updateOnGPS(String onGPS);
+    void updateOffGPS(String offGPS);
+    void updateSum(String sum);
+    void updateCtrList(CtrList ctrList);
+    void updateGeoList(GeoList geoList);
```
- 인터페이스 이름의 명확성에대해 조금더 고민해보면 좋을 것 같습니다! (의미 전달)
- 상태 변경이라도 반환값이 필요하진 않을지 한번은 고민해보면 좋을 것 같습니다!(이유 : 테스트 작성의 유리함 등..)

tomatozil replied:
아주 좋은 고민거리군요. Collection 구현체들의 메서드들을 생각해보면 void인 경우가 잘 없어요! 테스트 작성의 유리함 말고도 특별한 이유가 있을 것 같습니다.

Jeongjjuna commented on monicar-emulator/src/main/java/org/emulator/infrastructure/EmulatorConstant.java:
```diff
@@ -0,0 +1,8 @@
+    static final String TID = "A001";
+    static final String MID = "6";
+    static final String PV = "5";
+    static final String DID = "1";
```
유틸클래스에 final 붙여준것은 상속 불가능하기 위함이라고 보이는데요, 좋은것 같습니다!

정적 메소드나 필드를 모아놓고 사용하는 유틸리티 클래스의 경우는 객체를 만들 목적의 클래스가 아니기 때문에 명시적으로 priavate 생성자를 사용하는 것은 어떨까요?

tomatozil replied:
오호 잘 몰랐던 부분입니다.
찾아보니 인스턴스화 방지, 상속 방지를 위해 private 생성자를 만든다고 합니다. 
배워갑니다!

Jeongjjuna commented on monicar-emulator/src/main/java/org/emulator/infrastructure/InMemoryStore.java:
```diff
@@ -0,0 +1,32 @@
+public class InMemoryStore implements EmulatorRepository {
+    static Emulator emulator;
+
+    @Override
+    public void updateOnGPS(String onGPS) {
+        emulator.setOnGPS(onGPS);
+    }
+
+    @Override
+    public void updateOffGPS(String offGPS) {
+        emulator.setOffGPS(offGPS);
+    }
+
+    @Override
+    public void updateSum(String sum) {
+        emulator.setSum(sum);
+    }
```
setter보다 더 의미있는 메서드로 의미를 전달해주시면 좋을 것 같습니다!

Jeongjjuna reviewed (commented):
수정해보고 한번더 push 해주세요! 고생하셨습니다!

tomatozil committed:
Merge branch 'develop' into feature/#3-에뮬레이터-로컬-저장소-설계-및-구현2

tomatozil committed:
fix: 코드 리뷰 기반 수정

규격에 따른 dto와 비즈니스 로직에서 보다 쉽게 쓰기 위해 가독성을 향상시킨 infra class를 따로 뒀습니다
인스턴스화 방지, 상속 방지를 위해 유틸리티 클래스인 EmulatorConstant에 private 생성자를 추가했습니다
전역 데이터를 관리하는 클래스에 시동 on 정보와 off 정보를 누적하여 저장할 수 있도록 List를 추가했습니다

nailhae commented on monicar-emulator/src/main/java/org/emulator/infrastructure/VehicleEventData.java:
```diff
@@ -0,0 +1,17 @@
+package org.emulator.infrastructure;
+
+public class VehicleEventData {
+    private String carNumber;
```
전체적으로 차량에 vihicle을 사용하셨는데요.
여기와 `VehicleGlobalData`클래스에 `carNumber`대신 `vihicleNumber`로 사용하면 좀 더 일관성 있는 이름이 될 것 같아요.
[Pn룰](https://blog.banksalad.com/tech/banksalad-code-review-culture/#%EC%BB%A4%EB%AE%A4%EB%8B%88%EC%BC%80%EC%9D%B4%EC%85%98-%EB%B9%84%EC%9A%A9%EC%9D%84-%EC%A4%84%EC%9D%B4%EA%B8%B0-%EC%9C%84%ED%95%9C-pn-%EB%A3%B0)에서 P4정도 의견입니다!

--
P4: 반영해도 좋고 넘어가도 좋습니다 (Approve)

Jeongjjuna commented on monicar-emulator/src/main/java/org/emulator/infrastructure/VehicleGlobalData.java:
```diff
@@ -0,0 +1,79 @@
+        return data;
+    }
+
+    public Optional<VehicleEventData> getRecentVehicleStartData() {
+        if (vehicleStartDataList != null) {
+            return Optional.ofNullable(vehicleStartDataList.getLast());
+        }
+        return Optional.empty();
+    }
```
존재하지 않는 데이터 표현을 위해 옵셔널 활용.. 👍🏻

Jeongjjuna reviewed (approved):
고생하셨습니다!
추후 비즈니스 로직 개발하면서 더 개선해보면 좋을 것 같습니다! 👍🏻

nailhae commented on monicar-emulator/src/main/java/org/emulator/application/dto/OnOffDataDto.java:
```diff
@@ -0,0 +1,34 @@
+ * @param spd 속도 - 차량의 현재 속도
+ * @param sum 누적 주행 거리 - 차량의 현재 총 주행 거리
+ */
+public record OnOffDataDto(
```
`record`클래스라는 것이 있군요 배워갑니다!

tomatozil replied:
record는 class와 달리 데이터를 담고 운반하는 데에 초점을 맞춘 객체라고 합니다! getter를 만들어주고,, ㄷ등 우리가 해야할 작업들을 자동으로 처리해준다네요. dto와는 뭐가 다르냐? 불변객체여서 setter에 아예 열려있지 않습니다.
